### PR TITLE
Make sure we don't send messages multiple times in the availability d…

### PR DESCRIPTION
**Pull request by bkchr**
…istribution subsystem

### CODE REVIEW

File: node/network/availability-distribution/src/lib.rs [hunk 1]
```diff
@@ -364,24 +375,22 @@ where
...
 		let validator_index = obtain_our_validator_index(&validators, keystore.clone()).await;
 		state
-			.add_relay_parent(ctx, added, validators, validator_index)
+			.add_relay_parent(ctx, *added, validators, validator_index)
 			.await?;
 	}
 
 	// handle all candidates
-	for (candidate_hash, _receipt) in state.cached_live_candidates_unioned(added) {
+	for (candidate_hash, _receipt) in state.cached_live_candidates_unioned(view.difference(&old_view)) {
```
Comment (coriolinus) [inline node/network/availability-distribution/src/lib.rs:393]: So, previously we'd cached the result of `view.difference(&old_view)` as `added`. This PR removes that, and calculates the difference at least twice. `view.difference` requires `O(mn)` on the lengths of `view` and `old_view`. The implication of this change is that those sizes are small enough that recalculation is cheaper than allocation.

Does there exist profiling data to support that assertion?

**Review by coriolinus — APPROVED** [review]
`message_required_by_peer` looks good. It would be easier to review if the various formatting/non-semantic/optimization changes were in separate commits.

File: node/network/availability-distribution/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply node/network/availability-distribution/src/lib.rs:393]: In like 99% of the time there is `1` element in `view`.

**Review by ordian — APPROVED** [review]
Thanks!

**Comment by ordian** [comment]
bot merge

**Pull request merged**